certora/specs/atoken_balance.spec

```diff
@@ -0,0 +1,226 @@
+import "methods_base.spec"
+
+
+using AToken as _AToken 
+using DummyERC20_rewardToken as _DummyERC20_rewardToken
+using DummyERC20_aTokenUnderlying as _DummyERC20_aTokenUnderlying 
+
+
+methods
+{
+	// envfree
+	// -------
+	_AToken.balanceOf(address user) returns (uint256) envfree
+	_AToken.UNDERLYING_ASSET_ADDRESS() returns (address) envfree
+
+	// Permit
+    permit(address,address,uint256,uint256,uint8,bytes32,bytes32) => DISPATCHER(true)
+
+	// AToken
+	// ------
+    mint(address,address,uint256,uint256) returns (bool) => DISPATCHER(true)
+    burn(address,address,uint256,uint256) returns (bool) => DISPATCHER(true)
+    getIncentivesController() returns (address) => CONSTANT
+ }
+
+/**
+ * @title User AToken balance is fixed
+ * Interaction with `StaticAtokenLM` should not change a user's AToken balance,
+ * except for the following methods:
+ * - `withdraw`
+ * - `deposit`
+ * - `redeem`
+ * - `mint`
+ * - `metaDeposit`
+ * - `metaWithdraw`
+ *
+ * Note. Rewards methods are special cases handled in other rules below.
+ *
+ * Rules passed (with rule sanity): job-id=`5fdaf5eeaca249e584c2eef1d66d73c7`
+ *
+ * Note. `UNDERLYING_ASSET_ADDRESS()` was unresolved!
+ */
+rule aTokenBalanceIsFixed(method f) {
+	require _AToken == asset();
+	require _AToken.UNDERLYING_ASSET_ADDRESS() == _DummyERC20_aTokenUnderlying;
+	
+	// Limit f values
+	require (
+		(f.selector != deposit(uint256,address).selector) &&
+		(f.selector != deposit(uint256,address,uint16,bool).selector) &&
+		(f.selector != withdraw(uint256,address,address).selector) &&
+		(f.selector != redeem(uint256,address,address).selector) &&
+		(f.selector != redeem(uint256,address,address,bool).selector) &&
+		(f.selector != mint(uint256,address).selector) &&
+		(f.selector != metaDeposit(
+			address,address,uint256,uint16,bool,uint256,
+			(address,address,uint256,uint256,uint8,bytes32,bytes32),
+			(uint8, bytes32, bytes32)
+		).selector) &&
+		(f.selector != metaWithdraw(
+			address,address,uint256,uint256,bool,uint256,
+			(uint8, bytes32, bytes32)
+		).selector)
+	);
+
+	// Exclude reward related methods
+	require (
+		(f.selector != collectAndUpdateRewards(address).selector) &&
+		(f.selector != claimRewardsOnBehalf(address,address,address[]).selector) &&
+		(f.selector != claimSingleRewardOnBehalf(address,address,address).selector) &&
+		(f.selector != claimRewardsToSelf(address[]).selector) &&
+		(f.selector != claimRewards(address,address[]).selector)
+	);
+
+	env e;
+
+	// Limit sender
+	require e.msg.sender != currentContract;
+	require e.msg.sender != _AToken;
+
+	uint256 preBalance = _AToken.balanceOf(e.msg.sender);
+
+	calldataarg args;
+	f(e, args);
+
+	uint256 postBalance = _AToken.balanceOf(e.msg.sender);
+	assert preBalance == postBalance, "aToken balance changed by static interaction";
+}
+
+
```

Comment:
this can be interesting for modular approach - lets work on this @shoham-certora